### PR TITLE
[hail/ptypes] remove Subst from Emit, InferPTypes on ApplyComparisonOp

### PR DESCRIPTION
**Pull request by akotlar**

### CODE REVIEW

**Comment by tpoterba** [comment]
the 'subst' stands for 'substitute' -- I think that may have been a source of some of your confusion :)

**Comment by akotlar** [comment]
this was entirely autocorrect :(

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 1]
```diff
@@ -548,13 +548,15 @@ private class Emit(
         val vab = new StagedArrayBuilder(atyp.elementType, mb, 16)
         val sorter = new ArraySorter(er, vab)
 
-        val (array, compare, distinct) = (x: @unchecked) match {
-          case ArraySort(a, l, r, comp) => (a, Subst(comp, BindingEnv(Env[IR](l -> In(0, eltType), r -> In(1, eltType)))), Code._empty[Unit])
+        val (array, compare, distinct, fBindings: Array[String]) = (x: @unchecked) match {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Favor over Map, because the indices that were provided (into newMethod args), required knowledge by the caller, and an array is a clearer data structure for ordered collections (order here: left binding name, right binding name)

**Comment by tpoterba** [comment]
oh, ha.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: oh, yeah, totally.

**Comment by akotlar** [comment]
pytest and jvm-test pass

**Comment by akotlar** [comment]
> the 'subst' stands for 'substitute' -- I think that may have been a source of some of your confusion :)

I mostly got nervous explaining something new to me on the spot, and got lost in the details.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 2]
```diff
@@ -661,10 +667,10 @@ private class Emit(
             GetTupleElement(Ref("i-1", tt), tt.fields(0).index) -> GetTupleElement(Ref("i", tt), tt.fields(0).index)
         }
 
+        val compare2 = ApplyComparisonOp(EQWithNA(ktyp.virtualType), lastKey, currKey)
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: missed one infer, this isn't right however, need an env with these fields bound. fixing now.

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Fixed. Interesting: there's nothing that actually triggers this code in our test suite

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Actually python tests hit this; interesting that IRSuite.tetGroupKeyBy doesn't; looked briefly; exec strategy covers it, maybe gets optimized out?

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: added unoptimized compile testutil; did not apply to all jvmOnly, because a few tests fail (testScan, and one other). That is outside of the scope of this PR, but may be interesting.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Pretty sure this is getting constant-folded away.

Adding a no-opt compile path is a good change.

File: hail/src/main/scala/is/hail/expr/ir/Compile.scala [hunk 3]
```diff
@@ -124,6 +124,18 @@ object Compile {
     apply[AsmFunction5[Region, T0, Boolean, T1, Boolean, R], R](ctx, print, FastSeq((name0, typ0, classTag[T0]), (name1, typ1, classTag[T1])), body, optimize = true)
   }
 
+  def apply[T0: ClassTag, T1: ClassTag, R: TypeInfo : ClassTag](
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/Compile.scala]: where is this used?

File: hail/src/test/scala/is/hail/TestUtils.scala [hunk 4]
```diff
@@ -21,7 +21,7 @@ import org.apache.spark.sql.Row
 
 object ExecStrategy extends Enumeration {
   type ExecStrategy = Value
-  val Interpret, InterpretUnoptimized, JvmCompile, LoweredJVMCompile = Value
+  val Interpret, InterpretUnoptimized, JvmCompile, LoweredJVMCompile, JvmCompileUnoptimized = Value
```
Comment (tpoterba) [inline hail/src/test/scala/is/hail/TestUtils.scala]: we should add JVMCompileUnoptimized to the sets below, but let's do that in a separate PR -- have a feeling things are going to fail 😉

File: hail/src/main/scala/is/hail/expr/ir/Compile.scala (hunk 3 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/Compile.scala]: In testutils. JvmCompileUnoptimized path. Currently only enabled for testGroupByKey

File: hail/src/test/scala/is/hail/TestUtils.scala [hunk 5]
```diff
@@ -352,6 +354,16 @@ object TestUtils {
...
+            case ExecStrategy.JvmCompileUnoptimized =>
+              assert(Forall(x, node => node.isInstanceOf[IR] && Compilable(node.asInstanceOf[IR])))
+              eval(x, env, args, agg, bytecodePrinter =
+                Option(HailContext.getFlag("jvm_bytecode_dump"))
+                  .map { path =>
+                    val pw = new PrintWriter(new File(path))
+                    pw.print(s"/* JVM bytecode dump for IR:\n${Pretty(x)}\n */\n\n")
+                    pw
+                  },
+                optimize = false)
```
Comment (akotlar) [inline hail/src/test/scala/is/hail/TestUtils.scala]: Here

File: hail/src/test/scala/is/hail/expr/ir/IRSuite.scala [hunk 6]
```diff
@@ -2451,14 +2451,13 @@ class IRSuite extends HailSuite {
   }
 
   @Test def testGroupByKey() {
-    implicit val execStrats = ExecStrategy.javaOnly
+    implicit val execStrats = Set(ExecStrategy.Interpret, ExecStrategy.InterpretUnoptimized, ExecStrategy.JvmCompile, ExecStrategy.JvmCompileUnoptimized)
```
Comment (akotlar) [inline hail/src/test/scala/is/hail/expr/ir/IRSuite.scala]: And that branch is followed here

File: hail/src/main/scala/is/hail/expr/ir/Compile.scala (hunk 3 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Compile.scala]: can you point me to the line?

File: hail/src/test/scala/is/hail/TestUtils.scala (hunk 4 above)
Comment (akotlar) [reply hail/src/test/scala/is/hail/TestUtils.scala]: They definitely do, 2 failures at least, in one case a literal of an empty array, in the other (testCoalesce) possibly something similar

File: hail/src/main/scala/is/hail/expr/ir/Compile.scala (hunk 3 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/Compile.scala]: https://github.com/hail-is/hail/pull/8124#discussion_r381999923

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/Compile.scala]: I double checked that a constructor didn’t already exist. The combination of printer and optimize with 3 Long’s didn’t

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Compile.scala]: I can't see where you actually call compile in a changed line of code, though.

File: hail/src/test/scala/is/hail/TestUtils.scala (hunk 5 above)
Comment (tpoterba) [reply hail/src/test/scala/is/hail/TestUtils.scala]: This is a call to TestUtils.eval, not Compile?

Comment (akotlar) [reply hail/src/test/scala/is/hail/TestUtils.scala]: Compile is called in eval

Comment (akotlar) [reply hail/src/test/scala/is/hail/TestUtils.scala]: I believe

Comment (tpoterba) [reply hail/src/test/scala/is/hail/TestUtils.scala]: But eval worked before, and you didn't change eval.

File: hail/src/test/scala/is/hail/TestUtils.scala [hunk 7]
```diff
@@ -246,7 +248,7 @@ object TestUtils {
           val (resultType2, f) = Compile[Long, Long](ctx,
             argsVar, argsPType,
             MakeTuple.ordered(FastSeq(rewrite(Subst(x, BindingEnv(substEnv))))),
-            optimize = true,
+            optimize = optimize,
```
Comment (tpoterba) [inline hail/src/test/scala/is/hail/TestUtils.scala]: oh, I see! This is the change.

Comment (akotlar) [reply hail/src/test/scala/is/hail/TestUtils.scala]: 👍
